cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,327 @@
+                if log_level.value >= level.value:
+                    print(pad, log_level, *log_text)
+            print("\n", pad, "Final Circuit:", sep='')
+            print(textwrap.indent(str(log.final_circuit), pad))
+            print("----------------------------------------")
+
+        done = [0] * self._curr_id
+
+        def dfs(log: LoggerNode, pad=''):
+            print_log(log, pad)
+            done[log.transformer_id] = True
+            for child_id in log.nested_loggers:
+                dfs(self._logs[child_id], pad + ' ' * 4)
+
+        for i in range(self._curr_id):
+            if not done[i]:
+                dfs(self._logs[i])
+
+
+class NoOpTransformerStatsLogger(TransformerStatsLogger):
+    """All calls to this logger are a no-op"""
+
+    def register_initial(self, circuit: 'cirq.AbstractCircuit', transformer_name: str) -> None:
+        pass
+
+    def log(self, *args: str, level: LogLevel = LogLevel.INFO) -> None:
+        pass
+
+    def register_final(self, circuit: CIRCUIT_TYPE, transformer_name: str) -> None:
+        pass
+
+    def show(self, level: LogLevel = LogLevel.INFO) -> None:
+        pass
+
+
+@dataclasses.dataclass()
+class TransformerContext:
+    """Stores common configurable options for transformers.
+    Args:
+        logger: `TransformerStatsLogger` instance, which is a stateful logger used for logging the
+                actions of individual transformer stages. The same logger instance should be shared
+                across different transformer calls.
+        ignore_tags: Tuple of tags which should be ignored while applying transformations on a
+                circuit. Transformers should not transform any operation marked with a tag that
+                belongs to this tuple. Note that any instance of a Hashable type (like `str`,
+                `cirq.VirtualTag` etc.) is a valid tag.
+    """
+
+    logger: TransformerStatsLogger = NoOpTransformerStatsLogger()
+    ignore_tags: Tuple[Hashable, ...] = ()
+
+
+TRANSFORMER_TYPE = Callable[[CIRCUIT_TYPE, TransformerContext], CIRCUIT_TYPE]
+TRANSFORMER_TYPE_T = TypeVar(
+    'TRANSFORMER_TYPE_T',
+    TRANSFORMER_TYPE['cirq.Circuit'],
+    TRANSFORMER_TYPE['cirq.FrozenCircuit'],
+)
+TRANSFORMER_CLS_TYPE = TypeVar(
+    'TRANSFORMER_CLS_TYPE',
+    Type[TRANSFORMER_TYPE['cirq.FrozenCircuit']],
+    Type[TRANSFORMER_TYPE['cirq.Circuit']],
+)
+
+
+def _transform_and_log(
+    func: TRANSFORMER_TYPE[CIRCUIT_TYPE],
+    transformer_name: str,
+    circuit: CIRCUIT_TYPE,
+    context: TransformerContext,
+) -> CIRCUIT_TYPE:
+    """Helper to log initial and final circuits before and after calling the transformer."""
+
+    context.logger.register_initial(circuit, transformer_name)
+    circuit = func(circuit, context)
+    context.logger.register_final(circuit, transformer_name)
+    return circuit
+
+
+def transformer_method(func: TRANSFORMER_TYPE[CIRCUIT_TYPE]) -> TRANSFORMER_TYPE[CIRCUIT_TYPE]:
```

Comment:
The "method" terminology comes up in several places and should probably be changed to "function" or "func" throughout since this works for freestanding functions ("methods" are more specifically functions on classes).

Though I might suggest inlining the contents of this and `transformer_class` into `transformer` because I actually think having them close together makes more clear what's going on.

Reply:
inlined the contents